### ISSUE TITLE
Incorrect reference nullability warning when deconstructing result of the function

### ISSUE THREAD

**Issue by 13xforever** [dotnet/roslyn#49313]
**Version Used**: 
Dotnet SDK 5.0.0

**Steps to Reproduce**:
Not sure if relates to  #35349, but here's a repro:
```cs
using System;
using System.Threading.Tasks;

#nullable enable

public class C {
    public void M()
    {
        Result? r = null;
        r = r.NonNullableResult(); // result is marked as MayBeNull here
        (r, _) = r.MaybeNullableResult(); //warning CS8604: Possible null reference argument for parameter 'input' 
    }
}

public class Result
{
    public int Value {get; set;}
}

internal static class Extensions
{
    public static Result NonNullableResult(this Result? input)
        => input is null ? new() : input;
    
    public static (Result? result, bool extra) MaybeNullableResult(this Result input)
    {
        Result? tmp = input;
        try
        {
            input.Value = new Random().Next(42);
        }
        catch
        {
            tmp = null;
        }
        return (tmp, tmp is null);
    }
}
```

**Expected Behavior**:
No warning given

**Actual Behavior**:
I'd expect the result of `NonNullableResult()` to be not null, at least by the time `MaybeNullableResult()` is called on it.
if you change code to be
```cs
var tmp = r.MaybeNullableResult(); 
(r, _) = tmp;
```
warning goes away, which is kinda confirms it?

**Comment by jaredpar** [comment]
Think this repro can be simplified to the following 

```c#
using System;
using System.Threading.Tasks;

#nullable enable

public class C {
    static (C?, bool) M(C p) => throw null!;
    
    void Repro() {
         C? local = new C();
         
         // Warns
         (local, _) = M(local);
         
         // Does not warn
         local = new C();
         (_, _) = M(local);
    }
}
    
```

The core problem here is that we appear to be using the result of the deconstruction to update the state of the local before it is used in the call expression where it should be updated after.
